src/2025h2/autoreborrow-traits.md

```diff
@@ -0,0 +1,139 @@
+  slow.
+- Make sure autoreborrowing doesn't become a vehicle for implicit type coercion. Allowing autoreborrowing
+  from `T` to multiple values could be abused to define a `CustomInt(int128)` that coerces to all integer
+  types.
+  - Autoreborrow traits should use an associated type instead of a type parameter.
+  - Autoreborrowing at coercion sites should not dovetail into eg. an `Into::into` call.
+
+## Ownership and team asks
+
+| Task                         | Owner(s) or team(s) | Notes |
+|------------------------------|---------------------|-------|
+| Discussion and moral support | ![Team][] [lang]    | Normal RFC process |
+| Help around impl             | ![Team][] [compiler]| Trait-impl querying in rustc to replace `Pin<&mut T>` special case |
+| Do the work                  | @aapoalas           |       |
+
+### Design autoreborrow internals
+
+The basic idea of autoreborrowing is simple enough: when a reborrowable type is encountered at a coercion
+site, attempt a reborrow operation.
+
+Complications arise when reborrowing becomes recursive: if a `struct X { a: A, b: B }` contains two
+reborrowable types `A` and `B`, then we'd want the reborrow of `X` to be performed "piecewise". As an
+example, the following type should, upon reborrow, only invalidate any values that depend on the `'a` lifetime while any values dependent on the `'b` lifetime should still be usable as normal.
+
+```rust
+struct X<'a, 'b> {
+    a: &'a mut A,
+    b: &'b B,
+}
+```
+
+To enable this, reborrowing needs to be defined as a recursive operation but what the "bottom-case" is, that
+is the question. One option would be to use `!Copy + Reborrow` fields, another would use core marker types
+like `PhantomExclusive<'a>` and `PhantomShared<'b>` to discern the difference.
+
+
+| Task                 | Owner(s) or team(s)                | Notes                                                               |
+|----------------------|------------------------------------|---------------------------------------------------------------------|
+| Lang-team experiment | ![Team][] [lang]                   | allows coding pre-RFC; only for trusted contributors                |
+| Author RFC           | *Goal point of contact, typically* |                                                                     |
+| Lang-team champion   | ![Team][] [lang]                   | Username here |
+| RFC decision         | ![Team][] [lang]                   |                                                                     |
+| Secondary RFC review | ![Team][] [types]                  | request bandwidth from a second team, most features don't need this |
+
+### Implement non-recursive autoreborrowing
+
+A basic autoreborrowing feature should not be too complicated: the `Pin<&mut T>` special-case in the
+compiler already exists and could probably be reimagined to rely on a `Reborrow` trait.
```

Comment:
I'd be curious about any differences in behavior between the current implementation and the Reborrow one.